### PR TITLE
add getParametersTable for module

### PR DESCRIPTION
**Pull request by qiuxin2012**
## What changes were proposed in this pull request?

This pr provides a method to access parameters in a model by moduleName and parameterName.

## How was this patch tested?

unit tests

### CODE REVIEW

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 1]
```diff
@@ -280,7 +299,7 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
   override def hashCode(): Int = {
     def getHashCode(a: Object): Int = if (a == null) 0 else a.hashCode()
     val state = Seq(output, gradInput, this.getClass)
-    state.map(getHashCode).foldLeft(0)((a, b) => 31 * a + b)
+    state.map(getHashCode).foldLeft(0)((a, b) => 31 * a + b) + super.hashCode()
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: The breaks the define of hashCode.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala [hunk 2]
```diff
@@ -617,6 +617,16 @@ class BatchNormalization[@specialized(Float, Double) T: ClassTag](
     (Array(this.weight, this.bias), Array(this.gradWeight, this.gradBias))
   }
 
+  override def getParametersTable(): Table = {
+    if (affine) {
+      T(getName() -> T("weight" -> weight, "bias" -> bias,
+        "gradWeight" -> gradWeight, "gradBias" -> gradBias,
```
Comment (zhichao-li) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: gradWeight -> weightGrad, gradBias ->biasGrad

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Cosine.scala [hunk 3]
```diff
@@ -182,6 +183,10 @@ class Cosine[T: ClassTag](val inputSize : Int, val outputSize : Int)(
     (Array(this.weight), Array(this.gradWeight))
   }
 
+  override def getParametersTable(): Table = {
+    T(getName() -> T("weight" -> weight, "gradWeight" -> gradWeight))
```
Comment (zhichao-li) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Cosine.scala]: Move this as a default implementation to AbstractModule and only override it if needed.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Linear.scala [hunk 4]
```diff
@@ -158,6 +158,15 @@ class Linear[T: ClassTag](
     (Array(this.weight, this.bias), Array(this.gradWeight, this.gradBias))
   }
 
+  override def getParametersTable(): Table = {
```
Comment (zhichao-li) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Linear.scala]: can it be the general default implementation?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Cosine.scala (hunk 3 above)
Comment (qiuxin2012) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Cosine.scala]: 80% of the modules don't have weight and bias

File: dl/src/test/scala/com/intel/analytics/bigdl/nn/ModuleSpec.scala [hunk 5]
```diff
@@ -126,4 +127,40 @@ class ModuleSpec extends FlatSpec with Matchers {
     module2.eq(module) should be(false)
   }
 
+  "getName for different module" should "return different name" in {
+    val module1 = new Linear[Double](2, 3)
+    val module2 = new Linear[Double](4, 5)
+
+    module1.getName() should not be (module2.getName())
```
Comment (zhichao-li) [inline dl/src/test/scala/com/intel/analytics/bigdl/nn/ModuleSpec.scala]: add another checking: if the two models is equal than they share the same hashcode as we've just changed the hashcode method?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Cosine.scala (hunk 3 above)
Comment (zhichao-li) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Cosine.scala]: Or if there are no bias and no weight then we return None. Eliminate duplicate code if possible.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 6]
```diff
@@ -42,6 +42,7 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
 @specialized(Float, Double) T: ClassTag](
   implicit ev: TensorNumeric[T]) extends Serializable {
 
+  val UUID = java.util.UUID.randomUUID()
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: private?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Cosine.scala (hunk 3 above)
Comment (zhichao-li) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Cosine.scala]: @yiheng @qiuxin2012  maybe we can some how move weight or bias to an abstract class, then eliminate those duplicate *parameters() code in the following PR. 
What it want to get is simple, if bias is not null, add bias and if weight is not null add weight, but currently they are spread across many layers.

File: dl/src/test/scala/com/intel/analytics/bigdl/nn/ModuleSpec.scala (hunk 5 above)
Comment (qiuxin2012) [inline dl/src/test/scala/com/intel/analytics/bigdl/nn/ModuleSpec.scala]: Added

**Pull request merged**